### ISSUE TITLE
ConcurrentModificationException when running liquibase:diff

### ISSUE THREAD

**Issue by pkraeutli** [liquibase/liquibase-hibernate#75]
I get a java.util.ConcurrentModificationException when running the diff command against a PostgreSQL database. 

The command was working and the error only appeared after some time (probably after adding some specific mapping). I am running version 3.6-SNAPSHOT. Any ideas/pointers?

```
[ERROR] Failed to execute goal org.liquibase:liquibase-maven-plugin:3.3.1:diff (default-cli) on project xxx: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.liquibase:liquibase-maven-plugin:3.3.1:diff (default-cli) on project xxx: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	at org.codehaus.classworlds.Launcher.main(Launcher.java:46)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at org.liquibase.maven.plugins.AbstractLiquibaseMojo.execute(AbstractLiquibaseMojo.java:373)
	at org.liquibase.maven.plugins.LiquibaseDatabaseDiff.execute(LiquibaseDatabaseDiff.java:129)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 25 more
Caused by: liquibase.exception.LiquibaseException: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:123)
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:100)
	at org.liquibase.maven.plugins.LiquibaseDatabaseDiff.performLiquibaseTask(LiquibaseDatabaseDiff.java:149)
	at org.liquibase.maven.plugins.AbstractLiquibaseMojo.execute(AbstractLiquibaseMojo.java:369)
	... 28 more
Caused by: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at liquibase.command.AbstractCommand.execute(AbstractCommand.java:13)
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:121)
	... 31 more
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901)
	at java.util.ArrayList$Itr.next(ArrayList.java:851)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:235)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:221)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:241)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:241)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:241)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:263)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:60)
	at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:37)
	at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:25)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:126)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:119)
	at liquibase.command.DiffCommand.createReferenceSnapshot(DiffCommand.java:190)
	at liquibase.command.DiffCommand.createDiffResult(DiffCommand.java:140)
	at liquibase.command.DiffToChangeLogCommand.run(DiffToChangeLogCommand.java:51)
	at liquibase.command.AbstractCommand.execute(AbstractCommand.java:8)
	... 32 more
[ERROR] 
[ERROR] 
```

**Comment by msvticket** [comment]
I got the same error when diffing between postgres and hibernate. I did some debugging and found that the modification of fieldValue was done

	  at liquibase.ext.hibernate.snapshot.IndexSnapshotGenerator.snapshotObject(IndexSnapshotGenerator.java:41)
	  at liquibase.ext.hibernate.snapshot.HibernateSnapshotGenerator.snapshot(HibernateSnapshotGenerator.java:69)
	  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:50)
	  at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:163)

That is when snapshotId is null for an index include is called which changes fieldValue and you get this error when calling next on the iterator the next time.

**Comment by Jazzepi** [comment]
I got the same error.

Unexpected error running Liquibase: java.util.ConcurrentModificationException

SEVERE 3/15/15 4:41 PM: liquibase: java.util.ConcurrentModificationException
liquibase.exception.LiquibaseException: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
        at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:123)
        at liquibase.integration.commandline.Main.doMigration(Main.java:936)
        at liquibase.integration.commandline.Main.run(Main.java:175)
        at liquibase.integration.commandline.Main.main(Main.java:94)
Caused by: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
        at liquibase.command.AbstractCommand.execute(AbstractCommand.java:13)
        at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:121)
        ... 3 more
Caused by: java.util.ConcurrentModificationException
        at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:859)
        at java.util.ArrayList$Itr.next(ArrayList.java:831)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:235)
        at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:241)
        at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:263)
        at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
        at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
        at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:60)
        at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:37)
        at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:25)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:126)
        at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:119)
        at liquibase.command.DiffCommand.createReferenceSnapshot(DiffCommand.java:190)
        at liquibase.command.DiffCommand.createDiffResult(DiffCommand.java:140)
        at liquibase.command.DiffToChangeLogCommand.run(DiffToChangeLogCommand.java:51)
        at liquibase.command.AbstractCommand.execute(AbstractCommand.java:8)
        ... 4 more

When I added a @JoinTable to an @ElementCollection field.

    @ElementCollection(fetch = FetchType.LAZY)
    @OrderColumn
    @JoinTable(indexes = @Index(columnList = "textChunks"))
    @Fetch(FetchMode.JOIN)
    private List<String> textChunks = new ArrayList<>();

**Comment by todoesverso** [comment]
I git the same error doing diff between postgres and hibernate

**Comment by pauledler** [comment]
I got the same error doing diff between mysql & hibernate JPA's.  I am using version 3.5 of liquibase-hibernate4.  Is there any fix or workaround yet?

[ERROR] Failed to execute goal org.liquibase:liquibase-maven-plugin:3.3.2:diff (default-cli) on project aqtas-domain: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.liquibase:liquibase-maven-plugin:3.3.2:diff (default-cli) on project aqtas-domain: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error setting up or running Liquibase: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at org.liquibase.maven.plugins.AbstractLiquibaseMojo.execute(AbstractLiquibaseMojo.java:373)
	at org.liquibase.maven.plugins.LiquibaseDatabaseDiff.execute(LiquibaseDatabaseDiff.java:145)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 19 more
Caused by: liquibase.exception.LiquibaseException: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:123)
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:100)
	at org.liquibase.maven.plugins.LiquibaseDatabaseDiff.performLiquibaseTask(LiquibaseDatabaseDiff.java:176)
	at org.liquibase.maven.plugins.AbstractLiquibaseMojo.execute(AbstractLiquibaseMojo.java:369)
	... 22 more
Caused by: liquibase.command.CommandExecutionException: java.util.ConcurrentModificationException
	at liquibase.command.AbstractCommand.execute(AbstractCommand.java:13)
	at liquibase.integration.commandline.CommandLineUtils.doDiffToChangeLog(CommandLineUtils.java:121)
	... 25 more
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:859)
	at java.util.ArrayList$Itr.next(ArrayList.java:831)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:235)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:241)
	at liquibase.snapshot.DatabaseSnapshot.replaceObject(DatabaseSnapshot.java:263)
	at liquibase.snapshot.DatabaseSnapshot.includeNestedObjects(DatabaseSnapshot.java:195)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:177)
	at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:60)
	at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:37)
	at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:25)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:126)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:119)
	at liquibase.command.DiffCommand.createReferenceSnapshot(DiffCommand.java:190)
	at liquibase.command.DiffCommand.createDiffResult(DiffCommand.java:140)
	at liquibase.command.DiffToChangeLogCommand.run(DiffToChangeLogCommand.java:51)
	at liquibase.command.AbstractCommand.execute(AbstractCommand.java:8)
	... 26 more

**Comment by ledzpl** [comment]
Hello,
I got the same error doing diff between mysql 5 & hibernate JPA.
I am using liquibase-hibernate4(ver 3.5) and liquibase-core(ver 3.3.2).
There is no error when I add or remove fields but the error occurs when I add "@Index". 
I think It's so weird...To add @Index is very general works.. 

@Entity
@Table(name = "BOOK", indexes = { @Index(columnList = "title")  })
@Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
public class Book implements Serializable {
....
}

Is there any fix or workaround yet?

**Comment by dmitraver** [comment]
Hi!
Same for me actually. Is that planned to be fixed?

**Comment by benlucchesi** [comment]
I'm running jdk 8 - getting the same exception when adding indexes to a grails domain class.

any information on this issue?

**Comment by jkolobok** [comment]
Had the same problem. fixed by upgrading liquibase-core to 3.4.2

**Comment by nvoxland** [comment]
This should be fixed in the new Liquibase 3.5.1 + Hibernate 3.6 that should be released in the next week or so.
